File: src/c/ch01/1.3/l0/demo_properties.c
#include <stdio.h>
#include <stdint.h>
#include "parity.h"
#include "divisibility.h"
#include "remainder.h"

int main(void) {
    puts("Parity:");
    for (int64_t n : (int64_t[]){10, 7}) {
        printf("  %lld is_even? %s  is_odd? %s  (bit) %s\n",
               (long long)n,
               is_even(n) ? "true" : "false",
               is_odd(n)  ? "true" : "false",
               parity_bit(n));
    }

    puts("\nDivisibility:");
    printf("  12 divisible by 3? %s\n", is_divisible(12,3) ? "true" : "false");
    printf("  14 divisible by 5? %s\n", is_divisible(14,5) ? "true" : "false");

    puts("\nRemainders & identity:");
    int64_t q, r;
    div_identity(17, 5, &q, &r);
    printf("  17 = 5*%lld + %lld\n", (long long)q, (long long)r);

    puts("\nClock (7-day week):");
    /* 0=Mon, 5=Sat; Sat + 10 -> Mon(1) */
    printf("  Saturday(5) + 10 days -> %d\n", week_shift(5, 10));

    puts("\nLast digit of 2^15:");
    printf("  %lld\n", (long long)powmod(2, 15, 10));
    return 0;
}

File: archived/src/erlang/ch01/1.3/l0/src/demo_properties.erl
-module(demo_properties).
-export([main/0]).

main() ->
    io:format("Parity:~n", []),
    lists:foreach(
      fun(N) ->
          io:format("  ~p is_even? ~p  is_odd? ~p  (bit) ~s~n",
                    [N, parity:is_even(N), parity:is_odd(N), parity:parity_bit(N)])
      end,
      [10, 7]),

    io:format("~nDivisibility:~n", []),
    io:format("  12 divisible by 3? ~p~n", [divisibility:is_divisible(12,3)]),
    io:format("  14 divisible by 5? ~p~n", [divisibility:is_divisible(14,5)]),

    io:format("~nRemainders & identity:~n", []),
    {Q,R} = remainder:div_identity(17,5),
    io:format("  17 = 5*~p + ~p~n", [Q,R]),

    io:format("~nClock (7-day week):~n", []),
    io:format("  Saturday(5) + 10 days -> ~p~n", [remainder:week_shift(5,10)]),

    io:format("~nLast digit of 2^15:~n  ~p~n", [remainder:powmod(2,15,10)]).
